### PR TITLE
PRT-300 consumer state tracker

### PR DESCRIPTION
**Pull request by omerlavanet**
* **Please check if the PR fulfills these requirements**
- [ ] Commit message follows the Contribution Guidelines
- [ ] Tests ran locally and added/modified if needed
- [ ] Docs have been added/updated, if applicable
- [ ] If applicable - JIRA ticket ID was added

* **What kind of change does this PR introduce?** (Bug fix, feature, unit tests, docs update, ...)

* **What is the current behavior?** (You can also link to an open issue here)

* **What is the new behavior (if this is a feature change)?**

* **Please describe what manual tests you ran, if applicable**

* **Other information**:

### CODE REVIEW

File: protocol/rpcconsumer/rpcconsumer_server.go [hunk 1]
```diff
@@ -127,8 +128,11 @@ func (rpccs *RPCConsumerServer) SendRelay(
 	}
 
 	enabled, dataReliabilityThreshold := rpccs.chainParser.DataReliabilityParams()
+	fmt.Println("DataReliability enabled", enabled)
```
Comment (ranlavanet) [inline protocol/rpcconsumer/rpcconsumer_server.go:131]: remove debug prints

File: protocol/rpcconsumer/rpcconsumer_server.go [hunk 2]
```diff
@@ -127,8 +128,11 @@ func (rpccs *RPCConsumerServer) SendRelay(
...
+	fmt.Println("dataReliabilityThreshold", dataReliabilityThreshold)
 	if enabled {
 		for _, relayResult := range relayResults {
+			fmt.Println("USAOOO")
```
Comment (ranlavanet) [inline protocol/rpcconsumer/rpcconsumer_server.go:135]: remove debug prints here as well

File: protocol/rpcconsumer/rpcconsumer_server.go [hunk 3]
```diff
@@ -214,11 +218,13 @@ func (rpccs *RPCConsumerServer) sendRelayToProvider(
...
 	expectedBH, numOfProviders := rpccs.finalizationConsensus.ExpectedBlockHeight(rpccs.chainParser)
-	err = rpccs.consumerSessionManager.OnSessionDone(singleConsumerSession, epoch, reply.LatestBlock, chainMessage.GetServiceApi().ComputeUnits, relayLatency, expectedBH, numOfProviders, rpccs.consumerSessionManager.GetAtomicPairingAddressesLength()) // session done successfully
+	pairingAddressesLen := rpccs.consumerSessionManager.GetAtomicPairingAddressesLength()
+	latestBlock := relayResult.Reply.LatestBlock
+	err = rpccs.consumerSessionManager.OnSessionDone(singleConsumerSession, epoch, latestBlock, chainMessage.GetServiceApi().ComputeUnits, relayLatency, expectedBH, numOfProviders, pairingAddressesLen) // session done successfully
 
 	// set cache in a non blocking call
 	go func() {
-		err2 := rpccs.cache.SetEntry(ctx, relayRequest, chainMessage.GetInterface().Interface, nil, chainID, dappID, reply, relayResult.Finalized) // caching in the portal doesn't care about hashes
+		err2 := rpccs.cache.SetEntry(ctx, relayRequest, chainMessage.GetInterface().Interface, nil, chainID, dappID, relayResult.Reply, relayResult.Finalized) // caching in the portal doesn't care about hashes
```
Comment (ranlavanet) [inline protocol/rpcconsumer/rpcconsumer_server.go:227]: Using the same context in a go routine while returning the same context to the user might result in the context being canceled by the user. create a new context with a timeout from context.Background, especially for the caching.

File: protocol/rpcconsumer/rpcconsumer_server.go [hunk 4]
```diff
@@ -297,6 +303,7 @@ func (rpccs *RPCConsumerServer) sendDataReliabilityRelayIfApplicable(ctx context
 	// compare results for both relays, if there is a difference send a detection tx with both requests and both responses
 	specCategory := chainMessage.GetInterface().Category
 	if !specCategory.Deterministic || !relayResult.Finalized {
+		fmt.Println("IZASAOO", specCategory.Deterministic)
```
Comment (ranlavanet) [inline protocol/rpcconsumer/rpcconsumer_server.go:306]: remove debug prints, also who is IZASAOO ? 👍

File: protocol/rpcconsumer/rpcconsumer_server.go [hunk 5]
```diff
@@ -392,7 +399,10 @@ func (rpccs *RPCConsumerServer) sendDataReliabilityRelayIfApplicable(ctx context
 			report, conflicts := lavaprotocol.VerifyReliabilityResults(relayResult, dataReliabilityVerifications, numberOfReliabilitySessions)
 			if report {
 				for _, conflict := range conflicts {
-					rpccs.consumerTxSender.TxConflictDetection(ctx, nil, conflict, nil)
+					err := rpccs.consumerTxSender.TxConflictDetection(ctx, nil, conflict, nil)
+					if err != nil {
+						utils.LavaFormatError("could not send detection Transaction", err, nil)
```
Comment (ranlavanet) [inline protocol/rpcconsumer/rpcconsumer_server.go:404]: add some info from the conflict itself so we can debug?

File: protocol/statetracker/tx_sender.go [hunk 6]
```diff
@@ -5,17 +5,102 @@ import (
...
 
-func (ts *TxSender) New(ctx context.Context, txFactory tx.Factory, clientCtx client.Context) (ret *TxSender, err error) {
+func NewTxSender(ctx context.Context, clientCtx client.Context, txFactory tx.Factory) (ret *TxSender, err error) {
 	// set up the rpcClient, and factory necessary to make queries
+	clientCtx.SkipConfirm = true
+	ts := &TxSender{txFactory: txFactory, clientCtx: clientCtx}
+	return ts, nil
+}
+
+func (ts *TxSender) SimulateAndBroadCastTxWithRetryOnSeqMismatch(msg sdk.Msg) error {
```
Comment (ranlavanet) [inline protocol/statetracker/tx_sender.go:30]: Where is the retry?

Comment (ranlavanet) [reply protocol/statetracker/tx_sender.go:30]: Maybe add TODO:: add retry mechanism for sequence mismatch
 ;)

**Review by ranlavanet — CHANGES_REQUESTED** [review]

File: protocol/statetracker/tx_sender.go (hunk 6 above)
Comment (omerlavanet) [reply protocol/statetracker/tx_sender.go:30]: there is a TODO, and we will do it when we finish rpcprovider

**Review by ranlavanet — APPROVED** [review]

**Pull request merged**